arrow-core-data/src/main/kotlin/arrow/core/Ior.kt

```diff
@@ -492,8 +909,181 @@ fun <A> A.leftIor(): Ior<A, Nothing> = Ior.Left(this)
+
+fun <A, B> Ior<A, B>.replicate(SA: Semigroup<A>, n: Int): Ior<A, List<B>> =
+  if (n <= 0) Ior.Right(emptyList())
+  else when (this) {
+    is Ior.Right -> Ior.Right(List(n) { value })
+    is Ior.Left -> this
+    is Ior.Both -> bimap(
+      { List(n - 1) { leftValue}.fold(leftValue, { acc, a -> SA.run { acc + a }}) },
+      { List(n) { rightValue } }
+    )
+  }
+
+fun <A, B> Ior<A, B>.replicate(SA: Semigroup<A>, n: Int, MB: Monoid<B>): Ior<A, B> =
+  if (n <= 0) Ior.Right(MB.empty())
+  else when (this) {
+    is Ior.Right -> Ior.Right(MB.run { List(n) { value }.combineAll() })
+    is Ior.Left -> this
+    is Ior.Both -> bimap(
+      { List(n - 1) { leftValue}.fold(leftValue, { acc, a -> SA.run { acc + a }}) },
+      { MB.run { List(n) { rightValue }.combineAll() } }
+    )
+  }
+
+fun <A, B, C> Ior<A, Either<B, C>>.selectM(SA: Semigroup<A>, f: Ior<A, (B) -> C>): Ior<A, C> =
+  flatMap(SA) {
+    it.fold(
+      { b -> f.map { ff -> ff(b) } },
+      { c -> Ior.Right(c) }
+    )
+  }
+
+fun <A, B> Ior<A, Iterable<B>>.sequence(): List<Ior<A, B>> =
+  traverse(::identity)
+
+fun <A, B, C> Ior<A, Either<B, C>>.sequenceEither(): Either<B, Ior<A, C>> =
+  traverseEither(::identity)
+
+fun <A, B, C> Ior<A, Validated<B, C>>.sequenceValidated(): Validated<B, Ior<A, C>> =
+  traverseValidated(::identity)
 
-fun <E, A, B> Ior<E, A>.zip(SE: Semigroup<E>, fb: Ior<E, B>): Ior<E, Pair<A, B>> =
-  zip(SE, fb, ::Pair)
+/**
+ * Given [B] is a sub type of [C], re-type this value from Ior<A, B> to Ior<A, B>
+ *
+ * ```kotlin:ank:playground:extension
+ * import arrow.core.*
+ *
+ * fun main(args: Array<String>) {
+ *   //sampleStart
+ *   val string: Ior<Int, String> = Ior.Right("Hello")
+ *   val chars: Ior<Int, CharSequence> =
+ *     string.widen<Int, CharSequence, String>()
+ *   //sampleEnd
+ *   println(chars)
+ * }
+ * ```
+ */
+fun <A, C, B : C> Ior<A, B>.widen(): Ior<A, C> =
+  this
+
+fun <AA, A : AA, B> Ior<A, B>.leftWiden(): Ior<AA, B> =
+  this
+
+inline fun <A, B, C, Z> Ior<A, B>.zip(SA: Semigroup<A>, fb: Ior<A, C>, crossinline f: (B, C) -> Z): Ior<A, Z> =
+  ap(SA, fb.map { c: C -> { b: B -> f(b, c) } })
+
+fun <A, B, C> Ior<A, B>.zip(SA: Semigroup<A>, fb: Ior<A, C>): Ior<A, Pair<B, C>> =
+  zip(SA, fb, ::Pair)
+
+fun <A, B, C, Z> Ior<A, B>.zipEval(SA: Semigroup<A>, other: Eval<Ior<A, C>>, f: (B, C) -> Z): Eval<Ior<A, Z>> =
+  other.map {zip(SA, it).map { a -> f(a.first, a.second) }}
+
+fun <A, B> Semigroup.Companion.ior(SA: Semigroup<A>, SB: Semigroup<B>): Semigroup<Ior<A, B>> =
+  IorSemigroup(SA, SB)
+
+private class IorSemigroup<A, B>(
+  private val SGA: Semigroup<A>,
+  private val SGB: Semigroup<B>
+) : Semigroup<Ior<A, B>> {
+
+  override fun Ior<A, B>.combine(b: Ior<A, B>): Ior<A, B> =
+    combine(SGA, SGB, b)
+
+  override fun Ior<A, B>.maybeCombine(b: Ior<A, B>?): Ior<A, B> =
+    b?.let { combine(SGA, SGB, it) } ?: this
+}
+
+operator fun <A : Comparable<A>, B : Comparable<B>> Ior<A, B>.compareTo(other: Ior<A, B>): Int = fold(
```

Comment:
This cannot be implemented inside `Ior` since we cannot constraint `A` and `B` to `Comparable` there? So we can also not implemented `Comparable` for `Ior<A, B>`, correct?

Reply:
makes sense

Reply:
Yes, as we're adding an additional constraint for `A` and `B,` we cannot implement `compareTo` inside `Ior`. With this approach, `compareTo` will be available for `Ior` if and only if `A` and `B` implement `Comparable`. Does it make sense?